applications/spectrum_analyzer/spectrum_analyzer.c

```diff
@@ -0,0 +1,475 @@
+        break;
+    default:
+        min_hz = MIN_900 * 1000000;
+        max_hz = MAX_900 * 1000000;
+        break;
+    }
+
+    /* calibrate upper channels */
+    hz = freq * 1000000;
+    max_chan = NUM_CHANNELS / 2;
+    while (hz <= max_hz && max_chan < NUM_CHANNELS) {
+        model->chan_table[max_chan].frequency = hz;
+
+        FURI_LOG_T("Spectrum", "calibrate_freq ch[%u]: %lu", max_chan, hz);
+
+        hz += spacing;
+        max_chan++;
+    }
+
+    /* calibrate lower channels */
+    hz = freq * 1000000 - spacing;
+    min_chan = NUM_CHANNELS / 2;
+    while (hz >= min_hz && min_chan > 0) {
+        min_chan--;
+        model->chan_table[min_chan].frequency = hz;
+
+        FURI_LOG_T("Spectrum", "calibrate_freq ch[%u]: %lu", min_chan, hz);
+
+        hz -= spacing;
+    }
+
+    model->max_rssi = -200.0;
+    model->max_rssi_dec = 0;
+
+    FURI_LOG_D("Spectrum", "set_center_freq - max_hz: %u - min_hz: %u - spacing: %u Hz", max_hz, min_hz, spacing);
+    FURI_LOG_D("Spectrum", "ch[0]: %lu - ch[%u]: %lu", model->chan_table[0].frequency, NUM_CHANNELS - 1, model->chan_table[NUM_CHANNELS - 1].frequency);
+
+    // center_freq = freq;
+
+    set_filter();
+
+    return freq;
+}
+
+void spectrum_analyzer_thread(void* p) {
+    SpectrumAnalyzerContext* context = (SpectrumAnalyzerContext*)p;
+
+    // Start CC1101
+    furi_hal_subghz_reset();
+    furi_hal_subghz_load_preset(FuriHalSubGhzPresetOok650Async);
+    furi_hal_subghz_set_frequency(433920000);
+    furi_hal_subghz_flush_rx();
+    furi_hal_subghz_rx();
+
+    while(1) {
+        furi_hal_delay_ms(50);
+
+        SpectrumAnalyzerModel* model = (SpectrumAnalyzerModel*)acquire_mutex(context->model_mutex, 25);
+
+        // Thread hangs if using FURI_LOG_X...
+        // FURI_LOG_D("Spectrum", "spectrum_analyzer_thread: mutex acquired");
+
+        furi_hal_subghz_idle();
+        furi_hal_subghz_load_registers(radio_config);
+
+        uint8_t ch = 0; 
+
+        if (update_values_flag == true){
+            user_freq = set_center_freq(model, user_freq);
+            update_values_flag = false;
+        }
+
+        model->max_rssi_dec = 0;
+    
+        for (ch = 0; ch < NUM_CHANNELS-1; ch++) { 
+            // tune(ch);
+            furi_hal_subghz_set_frequency(model->chan_table[ch].frequency);
+
+            furi_hal_subghz_rx();
+            osDelay(3);
+            
+            //max_ss = 127 -> -10.5
+            //max_ss = 0 -> -74
+            //max_ss = 255 -> -74.5
+            //max_ss = 128 -> -138
+
+            model->chan_table[ch].ss = (furi_hal_subghz_get_rssi() + 138) * 2 ;
+
+            if (model->chan_table[ch].ss > model->max_rssi_dec) {
+                model->max_rssi_dec = model->chan_table[ch].ss;
+                model->max_rssi = (model->chan_table[ch].ss / 2) - 138;
+                model->max_rssi_channel = ch;
+            }
+
+            furi_hal_subghz_idle();
+        }  
+
+        release_mutex(context->model_mutex, model);
```

Comment:
Holding model for too long is going to lockup gui. 
Best option will be to generate data, then acquire mutex, then update, then release mutex.

Also there is a way to speedup work with transceiver: for time critical code acquire bus handle and then use cc1101 driver directly.

Reply:
Makes sense. 
Will research speeding up transceiver